src/oatpp/network/tcp/client/ConnectionProvider.cpp

```diff
@@ -124,32 +125,84 @@ std::shared_ptr<oatpp::data::stream::IOStream> ConnectionProvider::get() {
 #endif
 
   return std::make_shared<oatpp::network::tcp::Connection>(clientHandle);
+}
 
+std::shared_ptr<oatpp::data::stream::IOStream> ConnectionProvider::get(const std::chrono::duration<v_int64, std::micro>& timeout) {  
+  std::packaged_task<std::shared_ptr<oatpp::data::stream::IOStream>(network::Address)> task{&getConnection};
+  auto future = task.get_future();
+
+  if (timeout == std::chrono::microseconds::zero()) {
+      task(m_address);
+      return future.get();
+  }
+  
+  std::thread{std::move(task), m_address}.detach();
+  return future.wait_for(timeout) == std::future_status::ready ? future.get() : nullptr;
 }
 
-oatpp::async::CoroutineStarterForResult<const std::shared_ptr<oatpp::data::stream::IOStream>&> ConnectionProvider::getAsync() {
+oatpp::async::CoroutineStarterForResult<const std::shared_ptr<oatpp::data::stream::IOStream>&> ConnectionProvider::getAsync(const std::chrono::duration<v_int64, std::micro>& timeout) {
 
   class ConnectCoroutine : public oatpp::async::CoroutineWithResult<ConnectCoroutine, const std::shared_ptr<oatpp::data::stream::IOStream>&> {
   private:
     network::Address m_address;
     oatpp::v_io_handle m_clientHandle;
   private:
-    struct addrinfo* m_result;
+    struct AddrinfoWrapper {
+      struct addrinfo* m_result{nullptr};
+      
+      AddrinfoWrapper() = default;
+      AddrinfoWrapper(const AddrinfoWrapper&) = delete;
+      AddrinfoWrapper(AddrinfoWrapper&&) = delete;
+      ~AddrinfoWrapper() {
+        if (m_result != nullptr) {
+          freeaddrinfo(m_result);
+        }
+      }
+      AddrinfoWrapper& operator=(const AddrinfoWrapper&) = delete;
+      AddrinfoWrapper& operator=(AddrinfoWrapper&&) = delete;
+    };
+
+    std::shared_ptr<AddrinfoWrapper> m_resultWrapper{std::make_shared<AddrinfoWrapper>()};
     struct addrinfo* m_currentResult;
     bool m_isHandleOpened;
+
+    std::chrono::steady_clock::time_point m_startTime{std::chrono::steady_clock::now()};
+    std::chrono::duration<v_int64, std::micro> m_timeout;
+
+  private:
+    struct AddrinfoResult {
+      bool timedout;
+      bool successful;
+    };
+
+    AddrinfoResult getaddrinfoHelper(const oatpp::String& portStr, const struct addrinfo& hints) {
+      std::packaged_task<bool()> task{[=]() {
+        return 0 == getaddrinfo(m_address.host->c_str(), portStr->c_str(), &hints, &m_resultWrapper->m_result);
+      }};
+      auto future = task.get_future();
+
+      if (m_timeout == std::chrono::microseconds::zero()) {
+        task();
+        return {false, future.get()};
+      }
+
+      std::thread{std::move(task)}.detach();
```

Comment:
Unfortunately, we can't spawn threads from coroutines. An application can spawn millions of coroutines
but the max threads-count is very limited. If each coroutine spawns a thread - an application will crash.